README.md
[![travis](https://travis-ci.org/mapbox/mapbox-upload.svg?branch=master)](https://travis-ci.org/mapbox/mapbox-upload)

mapbox-upload
-------------
Library for interfacing with the Mapbox upload API.

## Authentication

A Mapbox API token is required for authentication. Generate a **secret** token
with the `uploads:write` scope enabled by following
[these steps](https://www.mapbox.com/help/create-api-access-token/).

## JavaScript Usage

```
$ npm install --save mapbox-upload
```

```javascript
var upload = require('mapbox-upload');

// creates a progress-stream object to track status of
// upload while upload continues in background
var progress = upload({
    file: __dirname + '/test.mbtiles', // Path to mbtiles file on disk.
    account: 'test', // Mapbox user account.
    accesstoken: 'validtoken', // A valid Mapbox API secret token with the uploads:write scope enabled.
    mapid: 'test.upload', // The identifier of the map to create or update.
    name: 'My upload' // Optional name to set, otherwise a default such as original.geojson will be used.
});

progress.on('error', function(err){
	if (err) throw err;
});

progress.on('progress', function(p){
	// Do something with progress-stream object, like display upload status
});

progress.once('finished', function(){
	// Upload has completed but is likely queued for processing and not yet available on Mapbox.
});

```

### Options

#### stream
A [`stream`](http://nodejs.org/api/stream.html) object can be passed in instead of `{file: filepath}`.
`length` option is recommended for accurate progress-stream reporting. If length is unknown, it can be updated after upload has begun by emitting a `length` event from the `stream` object.


## CLI Usage

Using the CLI will also require [generating a secret token](https://www.mapbox.com/help/create-api-access-token/) with the `uploads:write` scope enabled.

```
$ npm install --global mapbox-upload
$ export MapboxAccessToken=<access token with uploads:write scope enabled>
$ mapbox-upload username.dataid /path/to/file
```

CLI usage follows the following pattern:

```
mapbox-upload <dataset> [<filepath> | <url>]
```

- `dataset` refers to the id of the dataset or map being created or replaced.
- `file` or `url` refers to either:
  - a local file
  - a remote file on S3

**Creating** a new file might look like:

```
mapbox-upload <your-username>.create example.tif
```

**Updating** an existing file would look the same, except that you would pass
in an existing `id`.

## Tests

```
npm test
```

tests require env variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to generate test credentials


CHANGELOG.md
### v4.3.1

* Use the POST endpoint for upload credentials instead of GET

### v4.3.0

* :beetle: fix bug where the `--patch` flag wasn't be interpreted properly unless it was the last CLI argument, and added docs for the flags [#60](https://github.com/mapbox/mapbox-upload/issues/60)
* added `mapbox-upload@<version>` to the `POST` requests sent to the Uploads API [#30](https://github.com/mapbox/mapbox-upload/issues/30)
* Improved the error message for tokens without `uploads:write` scopes. Previously they returned a cryptic `Not Found` error message which wasn't very helpful [#27](https://github.com/mapbox/mapbox-upload/issues/27)

### v4.2.4

* validate name length before uploading [#64](https://github.com/mapbox/mapbox-upload/pull/64)

### v4.2.3

* allow spaces in names [#61](https://github.com/mapbox/mapbox-upload/pull/61)

### v4.2.2

* add regular expression to catch invalid characters for the `name` option ([#58](https://github.com/mapbox/mapbox-upload/issues/58))
* README updates
* support for Node v4.x

### v4.2.1

* replace s3-upload-stream with aws-sdk for multi-part uploads [#52](https://github.com/mapbox/mapbox-upload/pull/52)
* update default API url

### v4.2.0

* ensure all uploads are never cached and always unique [#47](https://github.com/mapbox/mapbox-upload/pull/47)

### v4.1.0

* add `name` flag and validation to pass in a custom name for uploads/tilesets [#45](https://github.com/mapbox/mapbox-upload/pull/45)

### v4.0.0

* update output to `tileset` per Mapbox API changes

### v3.2.0

* no sudo in travis tests
* improved CLI docs
* use console.log instead of util.print
* update dependency versions: https://github.com/mapbox/mapbox-upload/commit/7d7c40f78054e39517abab1670a8bb9fbcf07273

### v3.1.0

* enable `patch` mode [#22](https://github.com/mapbox/mapbox-upload/pull/22)
* improved tests [#26](https://github.com/mapbox/mapbox-upload/pull/26)
* only use `MapboxAPI` in the executable

### v3.0.0

* create CLI tool [#17](https://github.com/mapbox/mapbox-upload/pull/170)

### v2.0.0

* rename `putmap` to `createupload` to be in sync with the uploads API url change from `api/Map` to `v1/uploads`
* update `createupload` to work with any s3 url, not just the returned url [#16](https://github.com/mapbox/mapbox-upload/pull/16)
* refactor errors to fix some race conditions [#15](https://github.com/mapbox/mapbox-upload/pull/15)
* update aws-sdk to `2.0.29`

### v1.1.4

* update request dependency to `2.48.x`

### v1.1.3

* updated API url to `api.mapbox.com`
* no federation tokens in tests [#11](https://github.com/mapbox/mapbox-upload/pull/11)

### v1.1.2

* fix an improper error call [#8](https://github.com/mapbox/mapbox-upload/pull/8)

### v1.1.1

* update s3-upload-stream dependency to `0.6.x`

### v1.1.0

* handle callback errors
* coordinate with Mapbox APIs [#6](https://github.com/mapbox/mapbox-upload/pull/6)

### v1.0.0

* return a progress-stream synchronously

### v0.0.3

* improved error handling and better error messages

### v0.0.2

* allow more success status codes
* add TravisCI integration

### v0.0.1

* first


test_mpu.js
var fs = require('fs');
var stream = require('stream');
var AWS = require('aws-sdk');
var progress = require('progress-stream');
var mpuUploader = require('s3-upload-stream').Uploader;

var prog = progress({
        time: 100
    });

var st = fs.createReadStream('./test/test.mbtiles')
    .on('error', function(err) {
        upload.error(err, prog);
    });
    prog.setLength(fs.statSync('./test/test.mbtiles').size);

prog.on('progress', function(p){
    console.log('PROGRESS:', p)
});

var client = new AWS.S3({
    accessKeyId: process.env.AWS_KEY,
    secretAccessKey: process.env.AWS_SECRET,
    // sessionToken: ,
    region: "us-east-1"
});

// Set up read for file and start the upload.
var mpu = new mpuUploader({
        s3Client: client
    }, {
        Bucket: 'mapbox-upload-testing',
        Key: '_pending/test/badad68cc541a9b339565c1eb74d28cf' // Amazon S3 object name
    },
    // Callback handler
    function(err, uploadStream) {
        if (err) console.log(err);

        uploadStream.on('uploaded', function (data) {
            console.log('done', data);
        });

        uploadStream.on('chunk', function (data) {
            console.log('chunky', data);
        });
    st.pipe(prog).pipe(uploadStream)
    });

test/test.js
var test = require('tape');
var http = require('http');
var fs = require('fs');
var progress = require('progress-stream');
var request = require('request');
var upload = require(__dirname + '/../index.js');
upload.MAPBOX = 'http://localhost:3000';

var server;

function opts(extend) {
    extend = extend || {};
    var options = {
        file: __dirname + '/test.mbtiles',
        account: 'test',
        accesstoken: 'validtoken',
        mapid: 'test.upload'
    };
    for (var key in extend) options[key] = extend[key];
    return options;
};

test('setup', function(t) {
    function error(res, statusCode, message) {
        res.writeHead(statusCode);
        res.end(JSON.stringify({message:message}));
    }

    server = http.createServer(function(req, res) {
        switch (req.url) {
        case '/badjson/uploads/v1/test/credentials?access_token=validtoken':
            res.writeHead(200);
            res.end("hello world");
            break;
        case '/nokey/uploads/v1/test/credentials?access_token=validtoken':
            res.writeHead(200);
            res.end(JSON.stringify({bucket:'bar'}));
            break;
        case '/nobucket/uploads/v1/test/credentials?access_token=validtoken':
            res.writeHead(200);
            res.end(JSON.stringify({key:'bar'}));
            break;
        case '/uploads/v1/test/credentials?access_token=validtoken':
            upload.testcreds(function(err, data) {
                if (err) throw err;
                res.writeHead(200);
                res.end(JSON.stringify(data));
            });
            break;
        case '/uploads/v1/test?access_token=validtoken':
            if (req.method !== 'POST') return error(res, 404, 'Not Found');
            if (!req.headers['content-type'] ||
                req.headers['content-type'] !== 'application/json')
                return error(res, 400, 'Invalid content-type header');

            var body = '';
            req.on('data', function(chunk) {
                body += chunk.toString();
            });

            req.on('end', function() {
                try {
                    body = JSON.parse(body);
                } catch (e) {
                    return error(res, 400, 'Invalid JSON in body');
                }

                var schema = ['url', 'dataset'];
                for (var k in schema) if (!(schema[k] in body)) {
                    return error(res, 422, 'Missing property "' + schema[k] + '"');
                }
                for (var k in body) if (schema.indexOf(k) === -1) {
                    return error(res, 422, 'Invalid property "' + k + '"');
                }

                res.writeHead(201);
                res.end(JSON.stringify({
                    id: 'd51e4a022c4eda48ce6d1932fda36189',
                    progress: 0,
                    complete: false,
                    error: null,
                    created: '2014-11-27T18:47:30.065Z',
                    modified: '2014-11-27T18:47:30.065Z',
                    dataset: body.dataset,
                    owner: 'test'
                }));
            });
            break;
        case '/uploads/v1/test?access_token=invalid':
            if (req.method !== 'POST') return error(res, 404, 'Not found');
            error(res, 401, 'Unauthorized');
            break;
        case '/errorvalidjson/uploads/v1/test?access_token=validtoken':
            if (req.method !== 'POST') return error(res, 404, 'Not found');
            error(res, 400, 'Bad Request');
            break;
        case '/errorinvalidjson/uploads/v1/test?access_token=validtoken':
            if (req.method !== 'POST') return error(res, 404, 'Not found');
            res.writeHead(400);
            res.end('Bad Request');
            break;
        default:
            error(res, 404, 'Not found');
            break;
        }
    }).listen(3000);
    t.end()
});

test('upload.opts', function(t) {
    t.plan(6);
    t.throws(function() { upload.opts({}) }, /"file" or "stream" option required/);
    t.throws(function() { upload.opts({ file:'somepath' }) }, /"account" option required/);
    t.throws(function() { upload.opts({ file:'somepath', account:'test' }) }, /"accesstoken" option required/);
    t.throws(function() { upload.opts({ file:'somepath', account:'test', accesstoken:'validtoken' }) }, /"mapid" option required/);
    t.throws(function() { upload.opts({ file:'somepath', account:'test', accesstoken:'validtoken', mapid:'wrong.account' }) }, / Invalid mapid "wrong.account" for account "test"/);
    t.doesNotThrow(function() { upload.opts({ file:'somepath', account:'test', accesstoken:'validtoken', mapid:'test.upload' }) });
});

test('upload from file', function(t) {
    var prog = upload(opts());
    prog.on('error', function(err) {
        t.ifError(err);
    });
    prog.on('finished', function() {
        t.end();
    });
    prog.on('progress', function(p) {
        if (p.percentage === 100) {
            t.equal(100, p.percentage);
            t.equal(69632, p.length);
            t.equal(69632, p.transferred);
            t.equal(0, p.remaining);
            t.equal(0, p.eta);
        }
    })
});

test('upload from stream', function(t) {
    var options = opts();
    options.stream = fs.createReadStream(options.file)
    options.length = fs.statSync(options.file).size;
    options.file = null;
    var prog = upload(options);
    prog.on('error', function(err) {
        t.ifError(err);
    });
    prog.on('finished', function() {
        t.end();
    });
    prog.on('progress', function(p) {
        if (p.percentage === 100) {
            t.equal(100, p.percentage);
            t.equal(69632, p.length);
            t.equal(69632, p.transferred);
            t.equal(0, p.remaining);
            t.equal(0, p.eta);
        }
    })
});

test('upload.getcreds failed req', function(t) {
    upload.getcreds(opts({ mapbox: 'http://doesnotexist:9999' }), function cb(err, creds) {
        t.equal('getaddrinfo ENOTFOUND', err.message);
        t.end();
    });
});

test('upload.getcreds failed status', function(t) {
    upload.getcreds(opts({ mapbox: 'http://example.com' }), function cb(err, creds) {
        t.equal('Invalid JSON returned from Mapbox API: Unexpected token <', err.message);
        t.end();
    });
});

test('upload.getcreds failed badjson', function(t) {
    upload.getcreds(opts({ mapbox: 'http://localhost:3000/badjson' }), function cb(err, creds) {
        t.equal('Invalid JSON returned from Mapbox API: Unexpected token h', err.message);
        t.end();
    });
});

test('upload.getcreds failed no key', function(t) {
    upload.getcreds(opts({ mapbox: 'http://localhost:3000/nokey' }), function cb(err, creds) {
        t.equal('Invalid creds', err.message);
        t.end();
    });
});

test('upload.getcreds failed no bucket', function(t) {
    upload.getcreds(opts({ mapbox: 'http://localhost:3000/nobucket' }), function cb(err, creds) {
        t.equal('Invalid creds', err.message);
        t.end();
    });
});

test('upload.getcreds good creds', function(t) {
    upload.getcreds(opts(), function cb(err, c) {
        t.ifError(err);
        t.equal(c.bucket, 'mapbox-upload-testing');
        var keys = Object.keys(c);
        t.ok(keys.indexOf('bucket') > -1);
        t.ok(keys.indexOf('key') > -1);
        t.ok(keys.indexOf('accessKeyId') > -1);
        t.ok(keys.indexOf('secretAccessKey') > -1);
        t.end();
    });
});

test('upload.getcreds bad creds', function(t) {
    upload.getcreds(opts({ accesstoken: 'invalid' }), function cb(err) {
        t.equal(404, err.code);
        t.equal('Not found', err.message);
        t.end();
    });
});

test('upload.putfile failed no key', function(t) {
    function cb(err) {
        t.equal('"key" required in creds', err.message);
        t.end();
    };
    var prog = progress();
    prog.once('error', cb);
    upload.putfile(opts(), {}, prog);
});

test('upload.putfile failed no bucket', function(t) {
    function cb(err) {
        t.equal('"bucket" required in creds', err.message);
        t.end();
    };
    var prog = progress();
    prog.once('error', cb);
    upload.putfile(opts(), { key: '_pending' }, prog, cb);
});

test('upload.putfile good creds (file)', function(t) {
    upload.testcreds(function(err, creds) {
        t.ifError(err);
        function check() {
            request.head({
                uri: 'http://mapbox-upload-testing.s3.amazonaws.com/' + creds.key
            }, function(err, res, body) {
                t.ifError(err);
                t.equal(res.statusCode, 200);
                t.equal('69632', res.headers['content-length']);
                t.ok(+new Date(res.headers['last-modified']) > +new Date - 60e3);
                prog.called = true;
                t.end();
            });
        };
        var prog = progress();
        prog.on('finished', check);
        upload.putfile(opts(), creds, prog);
    });
});

test('upload.createupload good creds', function(t) {
    upload.testcreds(function(err, creds) {
        t.ifError(err);
        var url = 'http://' + creds.bucket + '.s3.amazonaws.com/' + creds.key;
        upload.createupload(url, opts(), function cb(err, body) {
            t.ifError(err);
            t.deepEqual(body, {
                id: 'd51e4a022c4eda48ce6d1932fda36189',
                progress: 0,
                complete: false,
                error: null,
                created: '2014-11-27T18:47:30.065Z',
                modified: '2014-11-27T18:47:30.065Z',
                dataset: 'test.upload',
                owner: 'test'
            });
            t.end();
        });
    });
});

test('upload.createupload bad creds', function(t) {
    upload.testcreds(function(err, creds) {
        t.ifError(err);
        var url = 'http://' + creds.bucket + '.s3.amazonaws.com/' + creds.key;
        upload.createupload(url, opts({accesstoken:'invalid'}), function cb(err) {
            t.equal(401, err.code);
            t.equal('Unauthorized', err.message);
            t.end();
        });
    });
});

test('upload.createupload error - valid json', function(t) {
    upload.testcreds(function(err, creds) {
        t.ifError(err);
        var url = 'http://' + creds.bucket + '.s3.amazonaws.com/' + creds.key;
        upload.createupload(url, opts({mapbox: 'http://localhost:3000/errorvalidjson'}), function cb(err) {
            t.equal(400, err.code);
            t.equal('Bad Request', err.message);
            t.end();
        });
    });
});

test('upload.createupload error - bad json', function(t) {
    upload.testcreds(function(err, creds) {
        t.ifError(err);
        var url = 'http://' + creds.bucket + '.s3.amazonaws.com/' + creds.key;
        upload.createupload(url, opts({mapbox: 'http://localhost:3000/errorinvalidjson'}), function cb(err) {
            t.equal(err.message, 'Invalid JSON returned from Mapbox API: Unexpected token B');
            t.end();
        });
    });
});

test('teardown', function(t) {
    server.close(t.end);
});


test/test.creds.js
// This policy allows upload to a single key in a testing bucket.
// All keys are removed from this object daily via a lifecycle rule.
module.exports = {
    bucket: 'mapbox-upload-testing',
    AWSAccessKeyId: 'AKIAIR36XSPJU7ZTK5MQ',
    acl: 'public-read',
    key: '_pending/test/badad68cc541a9b339565c1eb74d28cf',
    policy: 'eyJleHBpcmF0aW9uIjoiMjAyMy0wOC0wNVQyMDoyOTowNC44MjFaIiwiY29uZGl0aW9ucyI6W3siYnVja2V0IjoibWFwYm94LXVwbG9hZC10ZXN0aW5nIn0seyJhY2wiOiJwdWJsaWMtcmVhZCJ9LHsia2V5IjoiX3BlbmRpbmcvdGVzdC9iYWRhZDY4Y2M1NDFhOWIzMzk1NjVjMWViNzRkMjhjZiJ9LHsic3VjY2Vzc19hY3Rpb25fcmVkaXJlY3QiOiJodHRwOi8vbG9jYWxob3N0OjMwMDAvdGVzdC9jcmVhdGUvYmFkYWQ2OGNjNTQxYTliMzM5NTY1YzFlYjc0ZDI4Y2YifV19',
    signature: 'gLMlmBh5+ATlX5fjHHi/Hw5VPcA=',
    success_action_redirect: 'http://localhost:3000/test/create/badad68cc541a9b339565c1eb74d28cf',
    filename: '.mbtiles|.tilejson$'
};


index.js
var request = require('request');
var crypto = require('crypto');
var util = require('util');
var url = require('url');
var fs = require('fs');
var progress = require('progress-stream');
var AWS = require('aws-sdk');
var stream = require('stream');

module.exports = upload;

// Returns a progress stream immediately
function upload(opts) {
    var prog = progress({ time: 100 });

    opts = upload.opts(opts);

    upload.getcreds(opts, function(err, c) {
        if (err) return prog.emit('error', err);
        var creds = c;
        upload.putfile(opts, creds, prog);
    });
    return prog;
}
upload.MAPBOX = 'https://api.mapbox.com';

upload.opts = function(opts) {
    opts = opts || {};
    opts.proxy = opts.proxy || process.env.HTTP_PROXY;
    opts.mapbox = opts.mapbox || upload.MAPBOX;
    if (!opts.file && !opts.stream)
        throw new Error('"file" or "stream" option required');
    if (!opts.account)
        throw new Error('"account" option required');
    if (!opts.accesstoken)
        throw new Error('"accesstoken" option required');
    if (!opts.mapid)
        throw new Error('"mapid" option required');
    if (opts.mapid.split('.')[0] !== opts.account)
        throw new Error(util.format('Invalid mapid "%s" for account "%s"', opts.mapid, opts.account));
    if (! /^([\w-. ]+)$/.test(opts.name))
        throw new Error('"name" contains invalid characters');
    if (opts.name && opts.name.length > 64)
        throw new Error('"name" can only be 64 characters long');
    return opts;
};

upload.getcreds = function(opts, callback) {
    try { opts = upload.opts(opts) }
    catch(err) { return callback(err) }

    // Adds a randomized cache-busting query string. Credentials for an
    // upload should always be unique and never used from cache.
    crypto.pseudoRandomBytes(16, afterRandom);

    function afterRandom(err, decache) {
        if (err) return callback(err);
        decache = decache.toString('hex');
        request.post({
            uri: util.format('%s/uploads/v1/%s/credentials?access_token=%s&decache=%s', opts.mapbox, opts.account, opts.accesstoken, decache),
            headers: { 'Host': url.parse(opts.mapbox).host },
            proxy: opts.proxy
        }, afterRequest);
    }

    function afterRequest(err, resp, body) {
        if (err) return callback(err);
        try {
            body = JSON.parse(body);
        } catch(e) {
            err = new Error('Invalid JSON returned from Mapbox API: ' + e.message);
            return callback(err);
        }
        if (resp.statusCode !== 200) {
            if (body.message == 'Not Found') {
                err = new Error('Invalid access token. Make sure your token has the uploads:write scope.');
            } else {
                err = new Error(body && body.message || 'Mapbox is not available: ' + resp.statusCode);
            }
            err.code = resp.statusCode;
            return callback(err);
        }

        if (!body.key || !body.bucket) {
            return callback(new Error('Invalid creds'));
        } else if (resp.headers['x-cache'] === 'Hit from cloudfront') {
            return callback(new Error('Received cached credentials, retry upload'));
        } else {
            return callback(null, body);
        }
    }
};

upload.putfile = function(opts, creds, prog) {
    try { opts = upload.opts(opts) }
    catch(err) { return prog.emit('error', err) }

    if (!creds.key)
        return prog.emit('error', new Error('"key" required in creds'), prog);
    if (!creds.bucket)
        return prog.emit('error', new Error('"bucket" required in creds'), prog);

    var st;
    if (opts.stream) {
        if (!opts.stream instanceof stream) return prog.emit('error', new Error('"stream" must be an stream object'), prog);
        st = opts.stream;

        // if length isn't set progress-stream will not report progress
        if (opts.length) prog.setLength(opts.length);
        else st.on('length', prog.setLength);
    } else {
        if (!opts.file || typeof opts.file !== 'string') {
            return prog.emit('error', new Error('"file" must be an string'), prog);
        }
        st = fs.createReadStream(opts.file)
            .on('error', function(err) {
                prog.emit('error', err);
            });
        prog.setLength(fs.statSync(opts.file).size);
    }

    prog.on('progress', function(p){
        prog.emit('stats', p);
    });
    // Set up aws client
    var client = new AWS.S3({
        accessKeyId: creds.accessKeyId,
        secretAccessKey: creds.secretAccessKey,
        sessionToken: creds.sessionToken,
        region: 'us-east-1'
    });

    var params = {
        Bucket: creds.bucket,
        Key: creds.key,
        Body: st.pipe(prog)
    };
    var options = { partSize: 15 * 1024 * 1024, queueSize: 1 };
    var s3upload = client.upload(params, options).on('error', function(e){
        e = new Error(e || 'Upload to Mapbox.com failed');
        return prog.emit('error', e, prog);
    });

    s3upload.send(function() {
        var url = 'http://' + creds.bucket + '.s3.amazonaws.com/' + creds.key;
        upload.createupload(url, opts, function(err, body) {
            if (err) return prog.emit('error', err);
            prog.emit('finished', body);
        });
    });
};

upload.createupload = function(url, opts, callback) {
    try {
        opts = opts || {};
        opts.proxy = opts.proxy || process.env.HTTP_PROXY;
        opts.mapbox = opts.mapbox || upload.MAPBOX;
        if (!opts.account)
            throw new Error('"account" option required');
        if (!opts.accesstoken)
            throw new Error('"accesstoken" option required');
        if (!opts.mapid)
            throw new Error('"mapid" option required');
        if (opts.mapid.split('.')[0] !== opts.account)
            throw new Error(util.format('Invalid mapid "%s" for account "%s"', opts.mapid, opts.account));
    } catch(err) {
        return callback(err);
    }

    var uri = util.format('%s/uploads/v1/%s?access_token=%s', opts.mapbox, opts.account, opts.accesstoken);

    request.post({
        uri: uri,
        proxy: opts.proxy,
        headers: {
            'Content-type': 'application/json',
            'User-Agent': 'mapbox-upload@' + require(__dirname + '/package.json').version
        },
        body: JSON.stringify({
            url: url,
            tileset: opts.mapid,
            patch: opts.patch,
            name: opts.name
        })
    }, function(err, res, body) {
        if (err) return callback(err);
        try {
            body = JSON.parse(body);
        } catch(e) {
            err = new Error('Invalid JSON returned from Mapbox API: ' + e.message);
            return callback(err);
        }
        if (res.statusCode !== 201) {
            err = new Error(body && body.message || body);
            err.code = res.statusCode;
            return callback(err);
        }

        return callback(null, body);
    });
};

// Generate test-friendly upload credentials.
// Objects from the testing bucket are deleted via lifecycle rule daily.
upload.testcreds = function(callback) {
    var md5 = crypto.createHash('md5').update(Math.random().toString()).digest('hex');
    var key = '_pending/test/' + md5;

    if (!process.env.AWS_ACCESS_KEY_ID)
        return callback(new Error('env var AWS_ACCESS_KEY_ID required'));
    if (!process.env.AWS_SECRET_ACCESS_KEY)
        return callback(new Error('env var AWS_SECRET_ACCESS_KEY required'));

    callback(null, {
        bucket: 'mapbox-upload-testing',
        key: key,
        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
        sessionToken: process.env.AWS_SESSION_TOKEN
    });
};
